requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
+
+As a mitigation, if necessary, we would consider a buffer mechanism to reduce affected events to ~0.003%, we can provide further details on this solution, however we would ideally avoid implementing this as it increases the complexity of the system quite significantly.
+
+#### 2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+Example: Count of active users broken down by `paying` user property each month in 2021.
+
+After this change, if `paying` property of a user in Dec 2021 is changed to `false`, count of `paying` users does not change for e.g. March 2021. The  user will be considered non-paying until their `paying` property changes again2.
+
+It also means users could be counted as both `paying=true` and `paying=false` during a given month when they change state.
+
+User properties will change over time on events, some customers may rely on the existing behavior (where the latest property is assumed for all prior events), however the new behavior is considered much more accurate by our customer feedback.
+
+Our proposed mitigation or most use-cases (e.g. events for users who have paid at some point) cohorts can be used to achieve the same results, we will document this for affected users
+
+
+### Technical implications
+
+### Benefits
+1. Faster queries: remove the need to do JOINs for any insight queries.]
+2. Unlock: Our query results no longer change unless new data has come in - we make it possible to later implement time-series caching to speed up repeated queries over long time frames (>10x improvement expected on repeated calculations).
+4. We can sample by person_id in queries, which will make it possible to build a quick way to speed up queries in huge (1Bn+ events) installations with no additional query time resource cost.
+5. Queries will require less memory and no longer OOM (making it cheaper and more reliable to operate posthog)
+3. We can shard by person_id, simplifying the sharding story a lot and removing the need for cross-shard communication when sharding.
+
+Note that some of these changes have been benchmarked [here](https://github.com/PostHog/posthog/issues/7962)
+
+
+#### Challenges
+
+6. We will need to set up some strong consistency for message ordering in plugin-server and kafka, to make sure messages are processed in the right order per user.
+7. We will store more data in the events table. See below for more details.
+8. Deleting user properties becomes more expensive due to being stored in a redundant way.
```

Comment:
Do you see the migration as a potential challenge? How long did it take to set up the test table for us for example?

Reply:
Good question. This was also mentioned under open questions.

Data copying wise - no. As part of https://github.com/PostHog/posthog/issues/5684 I figured out a way to do the copying 20x faster than what we've done so far, which is good enough.

Infrastructure-wise - maybe. Due to the extra storage we might need to add nodes to our cluster, which has budget ramifications etc we'd need to figure out.

Implementation-wise - definetly! Actually ingesting data in a consistent way so events get up-to-date person properties is going to be a challenge due to similar reasons it's a challenge right now. :)

Reply:
Makes sense! I guess I was specifically worried about self hosting. I think there's a reasonable way to implement this whilst running both tables at the same time?

Reply:
Yeah, I think so:
1. We can leverage async migrations to do it in a non-intrusive way
2. We could add the person_id and person_properties columns to the existing table and backfill, but would need logic in the app to know when a query needs to join vs when it needs to read from the table. Some  questions around how to handle the difference in person semantics in the product will arise as a result but we can figure them out if this project kicks off :)

Reply:
It's an operation we currently provide via the UI and comes up relatively often around compliance requirements, hence brought up here. 

That said it's not in the critical path of the app and not a huge blocker when implementing it.

Reply:
I would argue it's pretty critical that we're able to properly delete user properties everywhere. Not only is it a huge compliance red flag, I think brand-wise it could hurt that we're the privacy friendly analytics, and you can't delete potentially highly sensitive data.

There also seems to be a non-trivial amount of [current usage](https://app.posthog.com/insights/MNop0r80/edit#filters=%7B%22insight%22%3A%22TRENDS%22%2C%22interval%22%3A%22month%22%2C%22display%22%3A%22ActionsLineGraph%22%2C%22actions%22%3A%5B%7B%22id%22%3A%2214240%22%2C%22name%22%3A%22person%20deleted%20click%22%2C%22type%22%3A%22actions%22%2C%22order%22%3A0%7D%5D%2C%22events%22%3A%5B%7B%22id%22%3A%22person%20viewed%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%2C%22name%22%3A%22person%20viewed%22%7D%5D%2C%22properties%22%3A%5B%5D%2C%22filter_test_accounts%22%3Afalse%2C%22new_entity%22%3A%5B%5D%2C%22date_from%22%3A%22-90d%22%2C%22date_to%22%3Anull%7D) (~1k persons deleted per month), plus we've seen cases from Enterprise customers where this becomes a total blocker.

Reply:
To make it clear, I'm not proposing dropping it: rather my comment was to point out it's not as performance-critical as say ingestion is and we can make this happen even in the new system by accepting the extra load it generates.